---
title: "Coverage-based identification of sex-linked contigs"
author: "sbai200"
date: "27/03/2022"
output: html_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```

# lcWGS

## Assess read quality

```{bash, eval = FALSE}
# FastQC v0.11.9 
# MultiQC v1.9 

# run on each of the fastq files
fastqc *.fastq*

# collate reports
multiqc .
```

## Trimmomatic

```{bash, eval = FALSE}
# Trimmomatic v0.39 
for infile in *_R1.fastq.gz
do
        base=$(basename ${infile} _R1.fastq.gz)
        trimmomatic PE -threads 8 ${infile} ${base}_R2.fastq.gz ${base}_R1_paired.T3PE2.fastq.gz ${base}_R1_unpaired.T3PE2.fastq.gz ${base}_R2_paired.T3PE2.fastq.gz ${base}_R2_unpaired.T3PE2.fastq.gz ILLUMINACLIP:/opt/nesi/CS400_centos7_bdw/Trimmomatic/0.39-Java-1.8.0_144/adapters/TruSeq3-PE-2.fa:2:30:10:2:keepBothReads LEADING:3 TRAILING:3 MINLEN:50

done
```

## fun fastqc + multiqc again to assess adaptor removal

## Map the reads from each individual to the draft genomes separately

Map to the male and female genomes (female genome alignments shown here as an example)

```{bash, eval = FALSE}
# BWA v0.7.17 

# Index the reference genome
bwa index <asm.fasta>

# Map the reads from each individual to the draft genomes separately


for infile in *_R1_paired.T3PE2.fastq
do
echo $infile

base=$(basename ${infile} _R1_paired.T3PE2.fastq)
echo $base

bwa mem -t 8 female_consensus_renamed ${infile} ${base}_R2_paired.T3PE2.fastq > ${base}_female.sam
done
```

## sort the alignments

```{bash, eval = FALSE}
# SAMtools v1.12

for infile in *_female.sam
do
echo $infile

base=$(basename ${infile} .sam)
echo "$base"

samtools sort -@ 8 -o ${base}_sorted.bam ${infile}
done
```

## Merge alignments

Generate 4 files: male reads female genome, female reads female genome, male reads male genome, female reads male genome. 
Female reads female genome are shown here as an example

## merge_females_female_genome

```{bash, eval = FALSE}
# SAMtools v1.12

# merge females mapped to the female genome

samtools merge merged_female_to_female_sorted.bam combined-1568-1-0_GGCTAC_female_sorted.bam combined-1568-2-0_CTTGTA_female_sorted.bam combined-1568-4-0_AGTTCC_female_sorted.bam combined-1568-5-0_ATGTCA_female_sorted.bam combined-1568-6-0_CCGTCC_female_sorted.bam combined-1568-8-0_GTGAAA_female_sorted.bam combined-1568-10-0_GTTTCG_female_sorted.bam

# from the ID's of the individuals we know which ones are male and female (3, 7 $ 9 are males)
```

## replace read groups

```{bash, eval = FALSE}
#GATK_4.1.8.1

#merged females to female genome
gatk AddOrReplaceReadGroups -I merged_female_to_female_sorted.bam -O merged_female_to_female_sorted.rg.bam  --RGLB WGS_females --RGPL Illumina --RGPU IA --RGSM WGS_females --RGID WGS_females  --CREATE_INDEX true
```

## Remove dups

```{bash, eval = FALSE}
gatk MarkDuplicates --REMOVE_DUPLICATES true -I merged_female_to_female_sorted.rg.bam  -O merged_female_to_female_sorted.rg.rmdup.bam -M merged_female_to_female.rmdup.metrics --CREATE_INDEX true --TMP_DIR tmp
```

# calculate genomic depth of coverage

Adapted from https://github.com/lurebgi/BOPsexChr/blob/master/m-f.coverage.sh

```{bash, eval = FALSE}
# Calculate per contig coverage
samtools depth -m 100 -q 20 -Q 10 merged_female_to_female_sorted.rg.rmdup.bam | awk '{print $1"\t"$2"\t"$2+1"\t"$3} ' > female_to_female.edited.depth

cut -f 1,2 Female_consensus_renamed.fai | awk '{print $1"\t1\t"$2}' > Female_consensus_renamed.fai.bed

bedtools map -a Female_consensus_renamed.fai.bed -b female_to_female.edited.depth -c 4 -o median,mean,count > female_to_female_contig.coverage

# Calculate coverage per 50Kb window
bedtools makewindows -g Female_consensus_renamed.fai -w 50000 > Female_consensus_renamed.windows.bed

bedtools map -a Female_consensus_renamed.windows.bed -b female_to_female.edited.depth -c 4 -o median,mean,count > female_to_female_50kb.coverage
```

# long-read (ONT) alignments

```{bash, eval=FALSE}
minimap2 -ax map-ont -t 8 Female_renamed_assembly Female_q10_reads > Mapped_female_reads_to_female_assembly.sam # align with minimap2

samtools sort -o Mapped_female_reads_to_female_assembly_sorted.bam -@ 16 Mapped_female_reads_to_female_assembly.sam # sort alignments

samtools view -b -F 256 Mapped_female_reads_to_female_assembly_sorted.bam > Mapped_female_reads_to_female_assembly_sorted_F.bam # remove secondary alignments

samtools index Mapped_female_reads_to_female_assembly_sorted_F.bam # index alignments

# calculate genomic depth of coverage as above
```

# R pipeline for whole contig cov

## Load packages

```{r}
library(tidyverse)
library(ggplot2)
library(magrittr)
```

## Import data

```{r}
female_columns <- c("Contig_id", "Female_start", "Female_end", "Female_median", "Female_mean", "Female_count")
male_columns <- c("Contig_id", "Male_start", "Male_end", "Male_median", "Male_mean", "Male_count")
types <- list("f", "i", "i", "i", "n", "i")
f2f_cov.df <- read_delim("female_to_female_contig.coverage", delim = " ", col_names = female_columns, col_types = types)

f2f_cov.df <- (f2f_cov.df[complete.cases(f2f_cov.df$Female_mean), ])

m2f_cov.df <- read_delim("male_to_female_contig.coverage", delim = " ", col_names = male_columns, col_types = types)

m2f_cov.df <- m2f_cov.df[complete.cases(m2f_cov.df[,5]), ]
```

## Normalise coverage

```{r}
correction_factor <- median(f2f_cov.df$Female_mean) / median(m2f_cov.df$Male_mean)

m2f_cov.df %<>% mutate(normalised_mean = Male_mean * correction_factor) 
```

## Join the male and female datasets

```{r}
Contig_cov.df <- left_join(f2f_cov.df, m2f_cov.df, by = "Contig_id")
Contig_cov.df[!complete.cases(Contig_cov.df),]
```

## Calculate the m:f cov

```{r}
Contig_cov.df %<>% mutate(m_f_ratio = normalised_mean / Female_mean)
```

## Calculate the ratio of male to female mappable sites

```{r}
Contig_cov.df %<>% mutate(m_map = Male_count / Female_end)
Contig_cov.df %<>% mutate(f_map = Female_count / Female_end)
Contig_cov.df %<>% mutate(m_f_map = m_map / f_map)
```

# R pipeline for 50 kb windows

## Load packages

```{r}
library(magrittr)
library(ggplot2)
library(tidyverse)
```

## load data

```{r}
# name columns
columns <- c("Contig_id", "Start", "End", "median", "mean", "count")
types <- list("f", "i", "i", "i", "n", "i")
# import data
cov_50kb.df <- read_delim("female_to_female_50kb.coverage", delim = " ", col_names = columns, col_types = types)

cov_50kb.df %>% group_by(Contig_id) %>% summarise(count = length(Contig_id))
```

## remove windows with no coverage

```{r}
no_cov <- cov_50kb.df[!complete.cases(cov_50kb.df$mean), ] # windows with no coverage
nrow(no_cov) # number of windows with no cov

cov_50kb.df <- cov_50kb.df[complete.cases(cov_50kb.df$mean), ] # remove no coverage windows
cov_50kb.df %>% group_by(Contig_id) %>% summarise(count = length(Contig_id)) # how many contigs
```

## get the lengths of the contigs

```{r}
names_fai <- c("Contig_id", "Length", "Offset", "Line_bases", "Line_width")
female_genome_fai.df <- read_delim("./../../../female_consensus_renamed.fasta - Copy.fai", delim = " ", col_names = names_fai)
```

## Remove small windows (smaller than 5Kb)

```{r}
cov_50kb.df %<>% mutate(window = End-Start) # Make windows

cov_50kb.df %<>% filter(window > 5000) # remove small windows

cov_50kb.df <- cov_50kb.df[complete.cases(cov_50kb.df$mean), ] # complete cases
```

## Remove the contigs that have less than 60% of the length covered by reads

```{r}
# get the sum of the count (the number of bases with coverage)
contig_lengths <- cov_50kb.df %>%
group_by(Contig_id) %>%
summarise(Sum_count = sum(count))

contig_lengths <- right_join(contig_lengths, female_genome_fai.df[,c(1,2)], by = "Contig_id") # add whole contig length
contig_lengths <- contig_lengths[complete.cases(contig_lengths), ] # remove empty rows
contig_lengths %<>% mutate(Percent_cov = Sum_count/Length*100) # get percentage of contig covered
contig_lengths %<>% filter(Percent_cov >= 60) # remove contigs with less than 60% of bases covered

# Join the coverage dataset to the filtered contig data
filtered_50kb_cov.df <- inner_join(cov_50kb.df, contig_lengths)
```

## plot coverage distribution

```{r}
filtered_50kb.cov.df %>% ggplot(aes(x = mean)) + geom_histogram(binwidth = 0.5, fill = '#21908CFF') + theme_classic() + ylab("Count of 50kb \n non-overlapping windows") + xlab("Coverage") 
```

## Split the alignments up into four groups low coverage, haploid coverage, diploid coverage and high coverage based on cutoffs in the histograms

```{r, warning=FALSE}
L = <low cutoff>
M = <middle cutoff>
H = <high cutoff>

# low coverage bins
low_cov.df <- filtered_cov_50kb.df.df %>% filter(mean <= L)

# half coverage bins
haploid.df <- filtered_cov_50kb.df.df %>% filter(mean > L) %>% filter(mean <= M)

# full coverage bins
diploid.df <- filtered_cov_50kb.df.df %>% filter(mean > M) %>% filter(mean <= H)

# high coverage bins
high_cov.df <- filtered_cov_50kb.df.df %>% filter(mean > H)
```

## calculate how many bp of each contig are in the low_cov group

```{r}
bases_low.df <- low_cov.df %>%
group_by(Contig_id) %>%
summarise(Bases_low = sum(count))

# add the full contig lengths to the dataframe
bases_low.df <- inner_join(bases_low.df, female_to_female_contig_lengths)

# add a new variable, % bases low
bases_low.df  %<>%  mutate(Percent_low = (Bases_low / Sum_count) *100)
```

## calculate how many bp of each contig are in the haploid group

```{r}
bases_hap.df <- haploid.df %>%
group_by(Contig_id) %>%
summarise(Bases_hap = sum(count))
# add the full contig lengths to the dataframe
bases_hap.df <- inner_join(bases_hap.df, female_to_female_contig_lengths)
# add a new variable, % bases hap
bases_hap.df  %<>%  mutate(Percent_hap = (Bases_hap / Sum_count) *100)
```

## calculate how many bp of each contig are in the diploid group

```{r}
bases_dip.df <- diploid.df %>%
group_by(Contig_id) %>%
summarise(Bases_dip = sum(count))
# add the full contig lengths to the dataframe
bases_dip.df <- inner_join(bases_dip.df, female_to_female_contig_lengths)
# add a new variable, % bases hap
bases_dip.df  %<>%  mutate(Percent_dip = (Bases_dip / Sum_count) *100)
```

## calculate how many bp of each contig are in the high_cov group

```{r}
bases_high.df <- high_cov.df %>%
group_by(Contig_id) %>%
summarise(Bases_high = sum(count))
# add the full contig lengths to the dataframe
bases_high.df <- inner_join(bases_high.df, female_to_female_contig_lengths)
# add a new variable, % bases hap
bases_high.df  %<>%  mutate(Percent_high = (Bases_high / Sum_count) *100)
```

## create a new dataframe containing the # bp of each contig per group and the contig lengths 

```{r}
low_and_hap.df <- full_join(bases_low.df, bases_hap.df)
dip_and_high.df <- full_join(bases_dip.df, bases_high.df)

new.df <- full_join(low_and_hap.df, dip_and_high.df)

# convert NA's into 0
new.df[is.na(new.df)] <- 0

# reorder columns
new.df <- new.df[, c(1, 2, 7, 9, 11, 6, 8, 10, 12, 4, 3, 5)]
```






