---
title: "Coverage-based identification of sex-linked contigs"
author: "sbai200"
date: "27/03/2022"
output: html_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```

# lcWGS

## Assess read quality

```{bash, eval = FALSE}
# FastQC v0.11.9 
# MultiQC v1.9 

# run on each of the fastq files
fastqc *.fastq*

# collate reports
multiqc .
```

## Trimmomatic

```{bash, eval = FALSE}
# Trimmomatic v0.39 
for infile in *_R1.fastq.gz
do
        base=$(basename ${infile} _R1.fastq.gz)
        trimmomatic PE -threads 8 ${infile} ${base}_R2.fastq.gz ${base}_R1_paired.T3PE2.fastq.gz ${base}_R1_unpaired.T3PE2.fastq.gz ${base}_R2_paired.T3PE2.fastq.gz ${base}_R2_unpaired.T3PE2.fastq.gz ILLUMINACLIP:/opt/nesi/CS400_centos7_bdw/Trimmomatic/0.39-Java-1.8.0_144/adapters/TruSeq3-PE-2.fa:2:30:10:2:keepBothReads LEADING:3 TRAILING:3 MINLEN:50

done
```

## fun fastqc + multiqc again to assess adaptor removal

## Map the reads from each individual to the draft genomes separately

Map to the male and female genomes (female genome alignments shown here as an example)

```{bash, eval = FALSE}
# BWA v0.7.17 

# Index the reference genome
bwa index <asm.fasta>

# Map the reads from each individual to the draft genomes separately


for infile in *_R1_paired.T3PE2.fastq
do
echo $infile

base=$(basename ${infile} _R1_paired.T3PE2.fastq)
echo $base

bwa mem -t 8 female_consensus_renamed ${infile} ${base}_R2_paired.T3PE2.fastq > ${base}_female.sam
done
```

## sort the alignments

```{bash, eval = FALSE}
# SAMtools v1.12

for infile in *_female.sam
do
echo $infile

base=$(basename ${infile} .sam)
echo "$base"

samtools sort -@ 8 -o ${base}_sorted.bam ${infile}
done
```

## Merge alignments

Generate 4 files: male reads female genome, female reads female genome, male reads male genome, female reads male genome. 
Female reads female genome are shown here as an example

## merge_females_female_genome

```{bash, eval = FALSE}
# SAMtools v1.12

# merge females mapped to the female genome

samtools merge merged_female_to_female_sorted.bam combined-1568-1-0_GGCTAC_female_sorted.bam combined-1568-2-0_CTTGTA_female_sorted.bam combined-1568-4-0_AGTTCC_female_sorted.bam combined-1568-5-0_ATGTCA_female_sorted.bam combined-1568-6-0_CCGTCC_female_sorted.bam combined-1568-8-0_GTGAAA_female_sorted.bam combined-1568-10-0_GTTTCG_female_sorted.bam

# from the ID's of the individuals we know which ones are male and female (3, 7 $ 9 are males)
```

## replace read groups

```{bash, eval = FALSE}
#GATK_4.1.8.1

#merged females to female genome
gatk AddOrReplaceReadGroups -I merged_female_to_female_sorted.bam -O merged_female_to_female_sorted.rg.bam  --RGLB WGS_females --RGPL Illumina --RGPU IA --RGSM WGS_females --RGID WGS_females  --CREATE_INDEX true
```

## Remove dups

```{bash, eval = FALSE}
gatk MarkDuplicates --REMOVE_DUPLICATES true -I merged_female_to_female_sorted.rg.bam  -O merged_female_to_female_sorted.rg.rmdup.bam -M merged_female_to_female.rmdup.metrics --CREATE_INDEX true --TMP_DIR tmp
```

# calculate genomic depth of coverage

Adapted from https://github.com/lurebgi/BOPsexChr/blob/master/m-f.coverage.sh

```{bash, eval = FALSE}
# Calculate per contig coverage
samtools depth -m 100 -q 20 -Q 10 merged_female_to_female_sorted.rg.rmdup.bam | awk '{print $1"\t"$2"\t"$2+1"\t"$3} ' > female_to_female.edited.depth

cut -f 1,2 Female_consensus_renamed.fai | awk '{print $1"\t1\t"$2}' > Female_consensus_renamed.fai.bed

bedtools map -a Female_consensus_renamed.fai.bed -b female_to_female.edited.depth -c 4 -o median,mean,count > female_to_female_contig.coverage

# Calculate coverage per 50Kb window
bedtools makewindows -g Female_consensus_renamed.fai -w 50000 > Female_consensus_renamed.windows.bed

bedtools map -a Female_consensus_renamed.windows.bed -b female_to_female.edited.depth -c 4 -o median,mean,count > female_to_female_50kb.coverage
```

R pipeline




